### ISSUE TITLE
[SCREENREADER]: Inline error messages are announcing to IE11 + JAWS when error appears and when it disappears

### ISSUE THREAD

**Issue by 1Copenut** [department-of-veterans-affairs/va.gov-team#314]
## Issue
Form inputs have inline error messages that are designed to announce themselves to screen readers when an input is invalid or empty. The error messages are firing correctly when inputs are blurred, but are announcing when users re-focus the input and provide a valid entry. The inputs are also not announcing the error message when users provide a second invalid entry.

This issue was reported and reproduced using **IE11 and JAWS** on Windows 7/10.

## Steps to Reproduce
1. Go to any form input, especially a standard `input[type="text"]` which is where the behavior was noted
2. Focus, and blur the input, preferably using `TAB` or arrow keys
3. Verify the error message announces itself
4. Refocus the input
5. Verify the error message is not read aloud
6. Key in a valid entry
7. Press `TAB` to leave (blur) the input
7. Verify the error message is read aloud, even though a valid entry was provided
8. Key in an invalid entry a second time
9. Blur and re-focus the input
10. Key in a second invalid entry
11. Verify the error message is not read aloud, even though the input was in error

## Related Issues
* https://github.com/usds/us-forms-system/issues/278

## Acceptance Criteria
- [ ] The error message should be announced to screen readers when an input is blurred with invalid or empty data. The error message should fire a second time when invalid data is entered a second time, but should not announce when users enter valid data to correct the first error.
- [ ] We should strongly consider adding the `aria-invalid="true"` attribute to our text inputs and text areas when an error is thrown. Some screen readers do not re-read the error message when inputs receive keyboard focus a second time. The `aria-invalid` attribute will announce to users the field under keyboard focus is invalid, and needs their attention.

## Possible Fix
I worked up a simple solution on CodePen, after much testing locally: https://codepen.io/tpierce_402/pen/Lqezye

I discovered in testing that creating a `<span>` to hold the error message and adding and removing the `role="alert"` attribute would stop IE11 + JAWS from reading the error message a second time. 

Visual users could have the same experience where the error doesn't drop out until they start typing. Aural users will hear **Field name, invalid text entry**. The error message will fire again when the field is blurred if the input is invalid or empty.

Tested on:
* Win10 + IE11 + JAWS
* Win10 + Chrome + JAWS
* Win10 + Firefox + NVDA
* MacOS Mojave + Safari + VoiceOver

---

## Screenshots
<img width="532" alt="screen shot 2018-09-20 at 9 54 57 pm" src="https://user-images.githubusercontent.com/934879/45858462-23631080-bd23-11e8-945a-9acd620487ad.png">

---

<img width="529" alt="screen shot 2018-09-20 at 9 55 13 pm" src="https://user-images.githubusercontent.com/934879/45858468-2a8a1e80-bd23-11e8-9a7e-20e4e7255641.png">

**Comment by andreahewitt-odd** [comment]
Moved to va.gov repo: https://app.zenhub.com/workspaces/vsp-5cedc9cce6e3335dc5a49fc4/issues/department-of-veterans-affairs/va.gov-team/1063

**Comment by peterhill-oddball** [comment]
> Moved to va.gov repo: https://app.zenhub.com/workspaces/vsp-5cedc9cce6e3335dc5a49fc4/issues/department-of-veterans-affairs/va.gov-team/1063

@andreahewitt-odd I think you mistakenly "migrated" this ticket.  But, it already exists in VA.GOV.  It does not need to be moved.  I did the same thing with two other issues :(

**Comment by 1Copenut** [comment]
@andreahewitt-odd && @peterhill-oddball I'm going to reopen this issue. It's something that affects IE11 + JAWS users.

**Comment by alexpappasoddball** [comment]
@brandonrapp looks like you unassigned me.  Do you want to update the labels to have it in your backlog (and remove the triage labels?)

**Comment by rianfowler** [comment]
the school feedback form would be a good place to recreate the issue and test your fix: https://staging.va.gov/education/submit-school-feedback/introduction. You can see this locally at the same path and it doesn't require sign in. 

this form is created using the [VA Forms Generator](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/forms). The config that drives it is [here](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/applications/edu-benefits/feedback-tool/config/form.js). You probably don't have to get into the nuts of bolts of the forms system to make this update. I think the issue probably requires and update to the ErrorableTextInput: 

- [documentation](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/visual-design/components/errorabletextinput/) 
- [code](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/applications/letters/components/ErrorableTextInput.jsx)

**Comment by rianfowler** [comment]
This [document](https://developer.paciellogroup.com/blog/2012/06/html5-accessibility-chops-aria-rolealert-browser-support/) outlines some approaches to this problem. We're currently using method 2 and trying out method 3. 

This is small but changes all of our forms app- because of this it could require refactoring several dozen unit tests and performing a lot of regression testing.

**Comment by 1Copenut** [comment]
@Clark-W-Griswold Adding you as an assignee for awareness.

**Comment by 1Copenut** [comment]
@rianfowler && @Clark-W-Griswold Bumping this one again for awareness. I had it tagged to the design system, but it's really focused on the forms library. This has been a long-running issue, so if there's any spare cycles, I'd love to chat or get this one added to a sprint. Thanks y'all.

**Comment by rianfowler** [comment]
@1Copenut should this be in ready? is there a pr that tries to fix it that puts it in the QA step?

**Comment by 1Copenut** [comment]
@rianfowler I don't believe there's a PR ready just yet. Can you point me to where this is in Ready state? I want to make sure I'm not accidentally putting issue tickets into the wrong bucket.

**Comment by rianfowler** [comment]
@1Copenut it's in the review/qa pipeline which I think is the last step before releasing 
![Screen Shot 2020-07-22 at 10 57 23](https://user-images.githubusercontent.com/4998130/88192093-245c7600-cc0a-11ea-9aaf-a94c593ec4ca.png)

**Comment by 1Copenut** [comment]
@rianfowler Yeah, this should def. be in `Ready` instead of `Review/QA`. There's been no work that I'm aware of. The fix that I discovered is a fairly complex rewrite, and I'm not sure how one would accomplish it in React. LMK if you want to discuss a strategy for the fix, or need help testing when work commences.

**Comment by caw310** [comment]
Need to retest on the minimum browsers but NOT IE11. 

Change severity to Defect 3.

**Comment by joshkimux** [comment]
Given this is specific to IE11, should we close this out @caw310 ?

**Comment by caw310** [comment]
Closing as this is old and the VADS has been updated. If there is still an issue file a new request.
